packages/orm/src/types.ts

```diff
@@ -11,6 +11,155 @@ import {
+  new (modelClass: typeof AnyModel, clauses?: QueryClause<Payload>[], opts?: object): QuerySet<MClassType>;
+}
+
+/**
+ * Handles relationships on the source model side.
+ *
+ * Each key points either to a single entity or a query set of entities on another model.
+ */
+export type TargetRelationship<
+  MClass extends AnyModel,
+  Relation extends Relations
+> = Relation extends Relations.OneToOne
+  ? SessionBoundModel<MClass>
+  : Relation extends Relations.ForeignKey
+    ? SessionBoundModel<MClass>
+    : Relation extends Relations.ManyToMany
+      ? QuerySet<ModelClassType<MClass>>
+      : never;
+      
+/**
+ * Handles relationships on the target model side.
+ *
+ * Defines keys used to access the Model foreign key is being defined from, from the target model.
+ */
+export type SourceRelationship<
+  MClassType extends typeof AnyModel,
+  Relation extends Relations
+> = Relation extends Relations.OneToOne
+  ? InstanceType<MClassType>
+  : Relation extends Relations.ForeignKey
+    ? QuerySet<MClassType>
+    : Relation extends Relations.ManyToMany
+      ? QuerySet<MClassType>
+      : never;
+      
+/**
+ * Possible types of relations and attributes that describe a single model.
+ */
+type ModelField = QuerySet<any> | AnyModel | Serializable;
+  
+/**
+ * A map of possible types of relations and attributes that describe a single model.
+ */
+export type ModelFieldMap = {
+  id?: ModelId;
+  [K: string]: ModelField;
+};
+  
+/**
+ * A plain JS object representing the database entry.
+ */
+export type Ref<MClass extends AnyModel> = ConstructorParameters<
+  ModelClassType<MClass>
+> extends [infer FirstConstructorParam]
+  ? FirstConstructorParam extends ModelRefLike
+  ? FirstConstructorParam
+  : never
+  : never;
+
+/**
+ * Extends the plan JS object interface, representing the database entry, with {@link ModelField}
+ * 
+ * Mainly used in functions that manipulate db entries.
+ */
+export type RefWithFields<M extends AnyModel> = {
+  [K in keyof ModelFields<M>]: ExcludeUndefined<ModelFields<M>[K]> extends QuerySet
+    ? (SessionBoundModel<ExcludeUndefined<ModelFields<M>[K]> extends QuerySet<infer MClass> ? InstanceType<MClass> : never> | ModelId | null)[]
+    : ExcludeUndefined<ModelFields<M>[K]> extends AnyModel
+      ? ModelFields<M>[K] | ModelId | null
+      : ModelFields<M>[K];
+}; 
+  
+/**
+ * Transforms the fields object to match the interface of the plain JS object in the database.
+ * 
+ * TODO: should firstly check if is undefined
+ */
+export type ModelRefLike<MFieldMap extends ModelFieldMap = ModelFieldMap> = {
+  [K in keyof MFieldMap]: ExcludeUndefined<MFieldMap[K]> extends QuerySet
+    ? never
+    : ExcludeUndefined<MFieldMap[K]> extends AnyModel
+      ? ModelId | undefined
+      : MFieldMap[K];
+};
+
+/**
+ * Excludes undefined type from an union of types.
+ */
+type ExcludeUndefined<T> = Exclude<T, undefined>;
```

Comment:
Excluding undefined from the union of types for types comparison (if X extends Y...). Maybe I could have get rid of this helper but IMO having this type does not harm the typing in some cases.